compose/config/config.py

```diff
@@ -238,114 +245,106 @@ def detect_cycle(self, name):
-        )
+        full_extended_config = pre_process_config(load_yaml(config_path))
 
         validate_extended_service_exists(
-            self.extended_service_name,
+            service_name,
             full_extended_config,
-            self.extended_config_path
+            config_path
         )
         validate_against_fields_schema(full_extended_config)
 
-        self.extended_config = full_extended_config[self.extended_service_name]
+        service_config = full_extended_config[service_name]
+        return config_path, service_config, service_name
 
-    def resolve_extends(self):
-        other_working_dir = os.path.dirname(self.extended_config_path)
+    def resolve_extends(self, extended_config_path, service_config, service_name):
+        other_working_dir = os.path.dirname(extended_config_path)
         other_already_seen = self.already_seen + [self.signature(self.service_name)]
 
         other_loader = ServiceLoader(
-            working_dir=other_working_dir,
-            filename=self.extended_config_path,
-            service_name=self.service_name,
-            service_dict=self.extended_config,
+            other_working_dir,
+            extended_config_path,
+            self.service_name,
+            service_config,
             already_seen=other_already_seen,
         )
 
-        other_loader.detect_cycle(self.extended_service_name)
+        other_loader.detect_cycle(service_name)
         other_service_dict = other_loader.make_service_dict()
         validate_extended_service_dict(
             other_service_dict,
-            filename=self.extended_config_path,
-            service=self.extended_service_name,
+            extended_config_path,
+            service_name,
         )
 
         return merge_service_dicts(other_service_dict, self.service_dict)
 
     def get_extended_config_path(self, extends_options):
-        """
-        Service we are extending either has a value for 'file' set, which we
+        """Service we are extending either has a value for 'file' set, which we
         need to obtain a full path too or we are extending from a service
         defined in our own file.
         """
         if 'file' in extends_options:
-            extends_from_filename = extends_options['file']
-            return expand_path(self.working_dir, extends_from_filename)
-
+            return expand_path(self.working_dir, extends_options['file'])
         return self.filename
 
     def signature(self, name):
-        return (self.filename, name)
+        return self.filename, name
+
+
+def resolve_environment(working_dir, service_dict):
+    """Unpack any environment variables from an env_file, if set.
+    Interpolate environment values if set.
+    """
+    if 'environment' not in service_dict and 'env_file' not in service_dict:
+        return {}
+
+    env = {}
+    if 'env_file' in service_dict:
+        for f in get_env_files(service_dict, working_dir=working_dir):
```

Comment:
Hah, I am not perfect either! While you're here making changes, may as well improve it :)